KEP for promoting AppArmor to GA

saschagrunert opened the pull request:
This is the first draft of promoting AppArmor to GA

Relates to: https://github.com/kubernetes/enhancements/issues/24
Analogous to: https://github.com/kubernetes/enhancements/pull/1148

/cc @timothysc

saschagrunert commented:
This is ready for a first rough review, if you have some cycles @kubernetes/sig-node-proposals

aojea commented on keps/sig-node/20200110-apparmor-ga.md:
```diff
@@ -0,0 +1,390 @@
+
+### Test Plan
+
+AppArmor already has [e2e tests][https://github.com/kubernetes/kubernetes/blob/6596a14/test/e2e_node/apparmor_test.go],
```
just curious, are those tests running as release blockers or release informers?

BenTheElder replied:
informing https://testgrid.k8s.io/sig-release-master-informing#node-kubelet-features

evrardjp commented:
Awesome to push that forward! I will have a read as soon as I can.

onlydole commented:
I'd love to help with the work defined in this KEP! I’ll keep my eyes on the threads 👀

saschagrunert commented:
Hey @timothysc, do you think we can target this KEP for 1.19? I think it would be a good fit and I can allocate some resources on my side to work on this.

BenTheElder commented:
re: https://github.com/kubernetes/enhancements/pull/1444#discussion_r365280097
> just curious, are those tests running as release blockers or release informers?

informing https://testgrid.k8s.io/sig-release-master-informing#node-kubelet-features

[moving my comment, it turns out if you have any existing review comments you are a reviewer forever, which puts the PR on gubernator.k8s.io/pr, I do not intend to review this PR currently]

tallclair commented:
The original plan was to wait to figure out how to get seccomp to GA before moving forward with this PR. However, there are a couple things that make Apparmor easier to move to GA:
1. The annotations are immutable (although the PSC annotations are not)
2. I don't think we need to deprecate localhost profiles

Given that, I think it's feasible to get this to GA for v1.19. I'm happy to be a reviewer, but I think you also need to get an api machinery API reviewer for some of the corner cases around the conversions. @liggitt was the reviewer on the seccomp to GA KEP, but I don't know whether he has time to take this on for v1.19

tallclair commented on keps/sig-node/20200110-apparmor-ga.md:
```diff
@@ -0,0 +1,390 @@
+This KEP proposes the absolute minimum to get AppArmor to GA, therefore all
+functional enhancements are out of scope, including:
+
+- Changing the default AppArmor profile from `unconfined`
```
The default apparmor profile is actually `runtime/default` so you can remove this.

saschagrunert replied:
Removed it as suggested.

tallclair commented on keps/sig-node/20200110-apparmor-ga.md:
```diff
@@ -0,0 +1,390 @@
+
+#### Pod API
+
+The Pod AppArmor API is generally immutable.
```
```suggestion
The Pod AppArmor API is generally immutable, except in `PodTemplates`.
```

saschagrunert replied:
Fixed

tallclair commented on keps/sig-node/20200110-apparmor-ga.md:
```diff
@@ -0,0 +1,390 @@
+
+type SecurityContext struct {
+    ...
+    // The AppArmor options to use by this container.
```
Add details about how this interacts with the pod options (copy from seccomp to GA)

saschagrunert replied:
Done :)

tallclair commented on keps/sig-node/20200110-apparmor-ga.md:
```diff
@@ -0,0 +1,390 @@
+type AppArmorOptions struct {
+    // The AppArmor profile to run with.
+    Profile
+}
```
We dropped this extra layer from the seccomp to GA proposal.

saschagrunert replied:
Removed that extra layer here as well.

tallclair commented on keps/sig-node/20200110-apparmor-ga.md:
```diff
@@ -0,0 +1,390 @@
+    // Use a predefined profile defined by the runtime.
+    // Most runtimes only support "default"
+    // +optional
+    RuntimeProfile *string
```
See seccomp proposal comments: specify format limits on this (length, valid characters, etc.)

saschagrunert replied:
Resolved by https://github.com/kubernetes/enhancements/pull/1444#discussion_r406069550

tallclair commented on keps/sig-node/20200110-apparmor-ga.md:
```diff
@@ -0,0 +1,390 @@
+    // Use a predefined profile defined by the runtime.
+    // Most runtimes only support "default"
+    // +optional
+    RuntimeProfile *string
```
I'm still torn on whether we should include this field initially. Currently there aren't any other runtime profiles accepted, and it makes reconciliation with the annotation more complicated.

I'm leaning towards dropping this field, and just having the type `Runtime` be synonymous with `runtime/default`

saschagrunert replied:
Alright, I removed the field and renamed the `Runtime` type to `Default`.

tallclair commented on keps/sig-node/20200110-apparmor-ga.md:
```diff
@@ -0,0 +1,390 @@
+    // Load a profile defined on the node.
+    // The profile must be available on the node to work.
+    // +optional
+    LocalhostProfile *string
```
What are the limitations on apparmor profile names in the apparmor spec? This field should be validated according to those.

saschagrunert replied:
Added the limitations and requirements to this field.

tallclair commented on keps/sig-node/20200110-apparmor-ga.md:
```diff
@@ -0,0 +1,390 @@
+
+The AppArmor fields on a pod are immutable.
+
+The behavior on annotation update is currently ill-defined: the annotation
```
Actually annotation updates for apparmor annotations are forbidden: https://github.com/kubernetes/kubernetes/blob/007f7ae7dc217d27dad994dce003cfe1c5bc6706/pkg/apis/core/validation/validation.go#L177

saschagrunert replied:
I removed that part of the sentence.

tallclair commented on keps/sig-node/20200110-apparmor-ga.md:
```diff
@@ -0,0 +1,390 @@
+fields.
+
+If both AppArmor annotations _and_ fields are specified, the values MUST match.
+This will be enforced in API validation.
```
Add a warning annotation (see seccomp proposal)

saschagrunert replied:
Added as suggested.

tallclair commented on keps/sig-node/20200110-apparmor-ga.md:
```diff
@@ -0,0 +1,390 @@
+API compatibility (no tightening validation), and makes a small stabilizing
+change to behavior (new Kubelets will ignore the update).
+
+#### PodSecurityPolicy Creation
```
Update the PSP sections to match the current seccomp proposal

saschagrunert replied:
Updated as suggested.

tallclair commented on keps/sig-node/20200110-apparmor-ga.md:
```diff
@@ -0,0 +1,390 @@
+#### PodTemplates
+
+PodTemplates (e.g. ReplaceSets, Deployments, StatefulSets, etc.) will be
+ignored. The field/annotation resolution will happen on template instantiation.
```
Add the warning here too (see seccomp proposal)

saschagrunert replied:
Added as suggested.

saschagrunert commented:
> The original plan was to wait to figure out how to get seccomp to GA before moving forward with this PR. However, there are a couple things that make Apparmor easier to move to GA:
> 
> 1. The annotations are immutable (although the PSC annotations are not)
> 2. I don't think we need to deprecate localhost profiles
> 
> Given that, I think it's feasible to get this to GA for v1.19. I'm happy to be a reviewer, but I think you also need to get an api machinery API reviewer for some of the corner cases around the conversions. @liggitt was the reviewer on the seccomp to GA KEP, but I don't know whether he has time to take this on for v1.19

Thank you for the review! I implemented most of your suggestions. @liggitt do you think we can move forward with this one?

liggitt commented:
I'm focusing on other GA efforts in 1.19 (CSR, Ingress, conformance without beta, deprecating beta APIs, and informing users and admins about deprecated API use) and don't have capacity to shepherd this for 1.19.

pjbgf commented on keps/sig-node/20200110-apparmor-ga.md:
```diff
@@ -0,0 +1,387 @@
+
+### Version Skew Strategy
+
+All API skew is resolved in the API server. New Kubelets will only use the
+AppArmor values specified in the fields and ignore the annotations therefore.
```
It may be worth reconsidering this. For seccomp the kubelet supports both, however, it prioritises fields over annotations. On that case, such support will then be dropped at v1.23.

If you were to change this, I would make sure you make that priority order explicit on this KEP.

saschagrunert replied:
Okay, changed the section to sync up with the seccomp KEP.

pjbgf commented on keps/sig-node/20200110-apparmor-ga.md:
````diff
@@ -0,0 +1,387 @@
+If both AppArmor annotations _and_ fields are specified, the values MUST match.
+This will be enforced in API validation.
+
+To raise awareness of annotation usage (in case of old automation), an
+additional warning annotation will be added when a pod is created with an
+AppArmor annotation:
+
+```
+warning.kubernetes.io/apparmor: "AppArmor set through annotations. Support will be dropped in v1.22"
+```
````
For seccomp we removed this specific approach and left a more generic entry, mentioning that a warning mechanism would be used, then highlighting which ones were being considered:

> The mechanisms being considerated are audit annotations, annotations on the object, events, or a 
> warning as described in [KEP #1693](/keps/sig-api-machinery/1693-warnings).

saschagrunert replied:
Thanks, changed that part as well.

saschagrunert commented on keps/sig-node/20200110-apparmor-ga.md:
```diff
@@ -0,0 +1,387 @@
+
+## Summary
+
+This is a proposal to upgrade the AppArmor annotation on pods & pod security
+policies (PSP) to a dedicated field, and mark the feature as GA. This proposal
```
Fixed, thanks! :pray:

saschagrunert commented on keps/sig-node/20200110-apparmor-ga.md:
```diff
@@ -0,0 +1,387 @@
+
+The main motivation behind this KEP is to promote the AppArmor feature to GA.
+
+_NOTE: Seccomp is in a very similar state, but with some subtle differences.
+Promoting Seccomp to GA will be covered by a [separate
+KEP][https://github.com/kubernetes/enhancements/pull/1148]._
```
Changed it :+1:

saschagrunert commented on keps/sig-node/20200110-apparmor-ga.md:
```diff
@@ -0,0 +1,387 @@
+### API
+
+The AppArmor API will be functionally equivalent to the current beta API. This
+includes the Pod API, which specifies what profile the containers run with, and
+the `PodSecurityPolicy` API which specifies allowed profiles & a default
+profile.
```
Added the note

saschagrunert commented on keps/sig-node/20200110-apparmor-ga.md:
````diff
@@ -0,0 +1,387 @@
+type PodSecurityPolicySpec struct {
+    ...
+    // AppArmor is the strategy that will dictate allowable and default AppArmor
+    // profiles for the container.
+    // +optional
+    AppArmor *ApparmorStrategyOptions
+    ...
+}
+
+type AppArmorStrategyOptions struct {
+    // The default profile to set on the pod, if none is specified.
+    // The default MUST be allowed by the allowedProfiles.
+    // +optional
+    DefaultProfile *v1.AppArmorProfile
+
+    // The set of profiles that may be set on the pod or containers.
+    // If unspecified, AppArmor profiles are unrestricted by this policy.
+    // +optional
+    AllowedProfiles *AppArmorProfileSet
+}
+
+// A set of AppArmor profiles. This struct should be a plural of
+// `v1.AppArmorProfile`.
+// All values are optional, and an unspecified field excludes all profiles of
+// that type from the set.
+type AppArmorProfileSet struct {
+    // The allowed AppArmor profile types.
+    // +optional
+    Types []AppArmorProfileType
+
+    // The allowed runtimeProfiles. A value of '*' allows all runtimeProfiles.
+    // +optional
+    RuntimeProfiles []string
+
+    // The allowed localhostProfiles. Values may end in '*' to include all
+    // localhostProfiles with a prefix.
+    // +optional
+    LocalhostProfiles []string
+}
+```
````
Moved it to the alternatives.

saschagrunert commented on keps/sig-node/20200110-apparmor-ga.md:
```diff
@@ -0,0 +1,387 @@
+API compatibility (no tightening validation), and makes a small stabilizing
+change to behavior (new Kubelets will ignore the update).
+
+#### PodSecurityPolicy Creation
+
+Unlike with pods, PodSecurityPolicy AppArmor annotations and fields are _not_
+synced.
+
+If only AppArmor annotations or fields are specified, no action is necessary.
+The set value is used when applying the PodSecurityPolicy.
+
+If both AppArmor annotations _and_ fields are specified, the values MUST match.
+This will be enforced in API validation.
+
+#### PodSecurityPolicy Update
+
+PodSecurityPolicy AppArmor fields are mutable. On an update, the same rules are
+applied as for creation, ignoring the old values.
+
+If only AppArmor annotations or fields are specified in the updated PSP, no
+action is necessary, and the specified values are used.
+
+If both AppArmor annotations _and_ fields are specified in the updated PSP, the
+values MUST match.
```
Yep, moved it down.

saschagrunert commented on keps/sig-node/20200110-apparmor-ga.md:
```diff
@@ -0,0 +1,387 @@
+Check these off as they are completed for the Release Team to track. These
+checklist items _must_ be updated for the enhancement to be released.
+
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP
```
How do we handle that for seccomp? Should I remove the PRR completely? I'm kinda confused now.

pjbgf replied:
There are a few sections around Production readiness review (PRR) that are new, which this KEP is missing. The PR I sent the link for shows what those new sections are.

For seccomp I just added those sections. Most of it did not apply or did not require much information.

saschagrunert commented:
The KEP has been updated to be marked as implementable for v1.20. Please take a look @tallclair @pjbgf @hasheddan

pjbgf reviewed (changes requested):
@saschagrunert thanks for working on this. There are some small nits, apart from that it is a /lgtm from me.

saschagrunert commented:
> @saschagrunert thanks for working on this. There are some small nits, apart from that it is a /lgtm from me.

Thank you for the review, I implemented all you suggestions! :pray: I just need a KEP number and I'm not sure where to get those.

Pinging @kubernetes/sig-node-pr-reviews for approval. :)

evrardjp reviewed (commented):
I don't know how this all works, but happy to see this tackled after the already good seccomp work :)

hasheddan commented on keps/sig-node/0000-apparmor-ga/README.md:
```diff
@@ -0,0 +1,685 @@
+
+- **What happens if we reenable the feature if it was previously rolled back?**
+
+  N/A - the feature is already enabled by default since Kubernetes 1.3.
```
Earlier in this document it mentions app armor was added in v1.4, which seems to align with [the docs](https://kubernetes.io/docs/tutorials/clusters/apparmor/). Not sure if maybe it was alpha for one release before becoming beta? Or should this also read 1.4?

saschagrunert replied:
Yep, I think 1.4 is correct. :+1: Fixed.

hasheddan commented on keps/sig-node/0000-apparmor-ga/README.md:
```diff
@@ -0,0 +1,685 @@
+  with and without the feature are necessary. At the very least, think about
+  conversion tests if API types are being modified.
+
+  N/A - the feature is already enabled by default since Kubernetes 1.3.
```
Same as comment above.

saschagrunert replied:
See https://github.com/kubernetes/enhancements/pull/1444#discussion_r478479559

saschagrunert commented:
Thank you for the review y'all! Fixed the commented parts.

evrardjp reviewed (commented):
lgtm so far, but someone else should approve. I am too new in here.

pjbgf reviewed (approved):
/lgtm

saschagrunert commented:
/assign @dchen1107 @derekwaynecarr

derekwaynecarr commented:
discussed in sig-node on 9/15, @tallclair to take over primary approval.

/unassign

tallclair commented on keps/sig-node/0000-apparmor-ga/README.md:
```diff
@@ -0,0 +1,685 @@
+| 1) Using custom or `runtime/default` profile when container runtime does not support AppArmor.     | Pod created          | The outcome is container runtime dependent. In this scenario containers may 1) fail to start or 2) run normally without having its policies enforced. |
+| 2) Using custom or `runtime/default` profile that restricts actions a container is trying to make. | Pod created          | The outcome is workload and AppArmor dependent. In this scenario containers may 1) fail to start, 2) misbehave or 3) log violations.                  |
+| 3) Using custom profile that does not exist on the node.                                           | Pod created          | Containers fail to start. Retry respecting RestartPolicy and back-off delay.                                                                          |
+| 4) Using a non supported runtime profile (i.e. `runtime/default-audit`).                           | Pod **not** created. | N/A                                                                                                                                                   |
```
nit
```suggestion
| 4) Using a unsupported runtime profile (i.e. `runtime/default-audit`).                           | Pod **not** created. | N/A                                                                                                                                                   |
```

tallclair commented on keps/sig-node/0000-apparmor-ga/README.md:
```diff
@@ -0,0 +1,685 @@
+
+| Scenario                                                                                           | API Server Result    | Kubelet Result                                                                                                                                        |
+| -------------------------------------------------------------------------------------------------- | -------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------- |
+| 1) Using custom or `runtime/default` profile when container runtime does not support AppArmor.     | Pod created          | The outcome is container runtime dependent. In this scenario containers may 1) fail to start or 2) run normally without having its policies enforced. |
```
There's another case to list here:

Host kernel or build does not have apparmor enabled - Kubelet will "soft fail" the pod by putting it in the `Blocked` state.

saschagrunert replied:
Added :+1:

tallclair commented on keps/sig-node/0000-apparmor-ga/README.md:
```diff
@@ -0,0 +1,685 @@
+### API
+
+The AppArmor API will be functionally equivalent to the current beta API. This
+includes the Pod API, which specifies what profile the containers run with, and
+the `PodSecurityPolicy` API which specifies allowed profiles & a default
+profile. There are no changes planned to the current `PodSecurityPolicy` API
```
nit: this is confusing with the following sentence. I'd just drop this.
```suggestion
includes the Pod API, which specifies what profile the containers run with. There are no changes planned to the current `PodSecurityPolicy` API
```

tallclair commented on keps/sig-node/0000-apparmor-ga/README.md:
```diff
@@ -0,0 +1,685 @@
+changes to the running containers that are using it, but will apply the updated
+profile on container restart. Note that amending such files is not recommended
+and may cause containers to fail on next restart, in the case of the new profile
+being more restrictive, invalid or the file no longer being present on disk.
```
AppArmor works differently from seccomp, and isn't file-based. Although the profiles are defined in files, to be used they must be loaded into the kernel (e.g. using the `apparmor-loader` program).

I'm actually not sure what happens if you reload the profile that's in use.

saschagrunert replied:
Updated the paragraph to reflect how profiles are loaded.

saschagrunert replied:
Did some research about the replacement and the AppArmor FAQ seems to answer this question:
https://gitlab.com/apparmor/apparmor/-/wikis/FAQ#what-is-profile-replacement

>  ### What is Profile Replacement?
>
> Profile replacement allows for replacing a currently loaded profile with a new version of the same profile. When this is done all tasks using the old profile begin using the new profile for all access decisions going forward. Doing this has several security implications that the administrator must be aware of:
> * Tasks retain access to all resources they had open under the older version of the profile
> * Any children tasks (applications a task has launched) retain the confinement that was decided by the old version of the profile. For children inheriting the parent profile (ix), then new children tasks will use the new profile.
> 
> The most secure solution is to always restart services after a profile has been updated, but in practice this is not always desired or required, which is why replacement is provided despite the above limitations

tallclair commented on keps/sig-node/0000-apparmor-ga/README.md:
```diff
@@ -0,0 +1,685 @@
+profiles across the cluster. Two examples being: 1) the lack of profile
+synchronization across nodes and 2) how difficult it can be to identify that
+profiles have been changed on disk, after pods started using it. However, given
+its current "pseudo-GA" state, we don't want to change it with this KEP.
```
The plan is to integrate it into the seccomp-operator, right? I think it's worth mentioning here (with the caveat that it's out of scope for this kep)

saschagrunert replied:
Yes, I mentioned it in another sentence :+1:

tallclair commented on keps/sig-node/0000-apparmor-ga/README.md:
```diff
@@ -0,0 +1,685 @@
+
+###### Profile files managed by the cluster admins
+
+The current support relies on profiles being saved as files in all cluster nodes
```
> ... being saved as files ...

As mentioned above, the requirement is that the profile is loaded into the kernel. This is typically done by writing it to a file first and then loading it.

saschagrunert replied:
Yes, I changed it to mention that the profiles have to be loaded.

tallclair commented on keps/sig-node/0000-apparmor-ga/README.md:
```diff
@@ -0,0 +1,685 @@
+
+Because the API is currently represented as (mutable) annotations, care must be
+taken for migrating to the API fields. The cases to consider are: pod create,
+pod update, PSP create, PSP update.
```
PSP create/update requirements were dropped, since we're not updating the PSP API

saschagrunert replied:
Fixed :+1:

tallclair commented on keps/sig-node/0000-apparmor-ga/README.md:
```diff
@@ -0,0 +1,685 @@
+
+When enforcing allowed profiles, the PSP should check BOTH the annotations &
+fields. In most cases, they should be consistent. On pod update, the AppArmor
+annotations may differ from the fields. In that case, the PSP enforcement should
```
Both the pod fields & annotations are immutable, so I don't think we need to worry about this

saschagrunert replied:
Yes, updated the section to reflect that.

tallclair commented on keps/sig-node/0000-apparmor-ga/README.md:
```diff
@@ -0,0 +1,685 @@
+The behavior on annotation update is currently ill-defined: the annotation
+update is allowed, but the new value will not be used until the container is
+restarted. There is no way to tell (from the API) what value a container is
+using.
```
Unlike seccomp, the apparmor annotations actually are immutable: https://github.com/kubernetes/kubernetes/blob/b46612a74224b0871a97dae819f5fb3a1763d0b9/pkg/apis/core/validation/validation.go#L177-L182

saschagrunert replied:
Ah thank you for the hint, I changed the paragraph to reflect that.

tallclair commented on keps/sig-node/0000-apparmor-ga/README.md:
```diff
@@ -0,0 +1,685 @@
+- No changes to the current behavior. Users are currently not allowed to specify
+  other runtime profiles. The existing API server rejects runtime profile names
+  that are different than `runtime/default`. The only exception being
+  `docker/default` - for backwards compatibility.
```
nit: `docker/default` is invalid for apparmor

saschagrunert replied:
Removed :+1:

tallclair reviewed (commented):
Mostly looks good. A handful of things that were copied over from the seccomp KEP don't apply here.

I also think we should remove the `AppArmor enabled` line from the node status: https://github.com/kubernetes/kubernetes/blob/cdbd07f64e2f1392fc9b3c68a6c1e06ee22d6449/pkg/kubelet/nodestatus/setters.go#L517-L523

tallclair commented on keps/sig-node/0000-apparmor-ga/README.md:
```diff
@@ -0,0 +1,682 @@
+AppArmor profiles are applied at container creation time. The underlying
+container runtime loads the profile referenced by its name via
+`apparmor_parser` on container creation. Therefore, updating the profiles
+content will only apply if the workload gets restarted.
```
IIRC, the runtime just uses profiles that have already been loaded by `apparmor_parser`, hence the need for the "apparmor-loader" image to load them: https://github.com/kubernetes/kubernetes/blob/69ae3144428e84069be928a8c70a356cb9bcb4b9/test/images/apparmor-loader/loader.go#L192

saschagrunert replied:
Yes, I rephrased this paragraph to mention that the runtime just references already loaded profiles. I also added a note that the profile have to be reloaded manually if changed on disk.

tallclair commented:
/lgtm

saschagrunert commented:
/assign @liggitt 
for (delegating or helping on) API review

@timothysc who would be the approver?

kikisdeliveryservice reviewed (commented):
@saschagrunert the dir structure seems to be incorrect: keps/sig-node/0000-apparmor-ga ---> keps/sig-node/24-apparmor-ga

saschagrunert commented:
> @saschagrunert the dir structure seems to be incorrect: keps/sig-node/0000-apparmor-ga ---> keps/sig-node/24-apparmor-ga

Thank you for the hint, fixed that :+1:

tallclair commented:
I believe @derekwaynecarr or @dchen1107 need to give the final approval, based on the OWNERS file.

liggitt commented on keps/sig-node/24-apparmor-ga/README.md:
```diff
@@ -0,0 +1,682 @@
+    // Load a profile defined on the node.
+    // The profile must be available on the node to work.
+    // The length of the profile is limited to 253 characters.
+    // The `Type` in this struct has to be set to `AppArmorProfileLocalhost`.
```
nit, put external-facing values in API doc, so `Localhost`

saschagrunert replied:
Fixed as suggested :+1:

liggitt commented on keps/sig-node/24-apparmor-ga/README.md:
```diff
@@ -0,0 +1,682 @@
+
+- **How can this feature be enabled / disabled in a live cluster?**
+
+  This feature was implemented prior to the concept of feature gates. Meaning,
+  it is already enabled in live clusters and cannot be disabled. This KEP is
+  simply "cleaning up" the API to make it GA.
```
is this accurate? disabling the AppArmor feature drops the annotations when creating a pod today - see https://github.com/kubernetes/kubernetes/blob/7705fa295c6470a3e9b34ddad4f5304ed8058d88/pkg/api/pod/util.go#L364-L370

saschagrunert replied:
Ah yeah, looks like it was one of the first features…changed the paragraph to:
```
  The feature can be enabled/disabled by the `AppArmor` feature gate. This
  behavior will not change when moving to GA.
```

liggitt commented on keps/sig-node/24-apparmor-ga/README.md:
```diff
@@ -0,0 +1,682 @@
+- **Can the feature be disabled once it has been enabled (i.e. can we rollback
+  the enablement)?**
+
+  This feature (AppArmor) cannot be disabled, as it is currently in a "pseudo-GA"
+  state. However, the changes it brings are backwards compatible, and the API
```
I'm not sure this is accurate, see https://github.com/kubernetes/kubernetes/blob/7705fa295c6470a3e9b34ddad4f5304ed8058d88/pkg/api/pod/util.go#L364-L370

saschagrunert replied:
Yes, that's clearly wrong. Fixed the paragraph.

liggitt commented on keps/sig-node/24-apparmor-ga/README.md:
```diff
@@ -0,0 +1,682 @@
+    ...
+    // The AppArmor options to use by the containers in this pod.
+    // +optional
+    AppArmor  *AppArmorProfile
```
pod-level apparmor profile is new to this KEP... the annotation-based feature only allows specifying container-level, correct?

saschagrunert replied:
Yes that's correct. My idea was to apply the same hierarchy like we have it right now for seccomp. Basically if a profile gets added to a sandbox, it automatically applies to all containers. Containers security context have a higher priority than on the pod level. Do you think this is implementable? We do not pass the profile on sandbox creation via the CRI, so it would be just a pod to container field copy.

liggitt replied:
I don't have enough context to know how tailored to a particular container AppArmor profiles typically are... is a pod-wide AppArmor profile likely to work with arbitrary injected sidecar containers?

liggitt replied:
On the other hand, someone might prefer an injected sidecar picking up the pod-wide default and failing over running unconfined? Not sure...

saschagrunert replied:
> I don't have enough context to know how tailored to a particular container AppArmor profiles typically are... is a pod-wide AppArmor profile likely to work with arbitrary injected sidecar containers?

At least the runtime/default profile (originated from docker) can be used with a wide variety of applications. From my viewpoint having a pod-wide field would be beneficial.

pjbgf replied:
I believe AppArmor's usage tend to be quite analogous to seccomp, although the former can be a lot more detailed, in both cases having profiles set at both Pod and Container levels would be quite useful, specially when considering multi-container pods.

For arbitrarily injected sidecars this becomes a bit of an interesting problem, as the pod level profile would need to account for its own workload plus an external workload that it would not know until deployment time. But on that case, the container-level profile override would resolve the issue, which is the same situation we have for seccomp.

tallclair replied:
Yeah, I suspect with both AppArmor & seccomp the usecase for a pod-level profile is something akin to the default profile, with fairly broad grants. Application-specific profiles, will likely need to be tied to the container. With seccomp, the pod-level field seems slightly more useful since the "runtime default" profile is not actually the default, but I also agree that the APIs should probably mirror each other.

liggitt commented on keps/sig-node/24-apparmor-ga/README.md:
```diff
@@ -0,0 +1,682 @@
+#### Pod Creation
+
+If no AppArmor annotations or fields are specified, no action is necessary.
+
```
Describe the behavior if the AppArmor feature (currently beta and enabled by default) is disabled. Currently the annotations are cleared. I would expect the new fields to behave the same as the annotations.

saschagrunert replied:
Yes, I added this:
```
If the `AppArmor` feature is disabled per feature gate, then the annotations are
cleared (current behavior) and the security context fields are being ignored.
```

liggitt commented on keps/sig-node/24-apparmor-ga/README.md:
```diff
@@ -0,0 +1,682 @@
+1. Container-specific field.
+2. Container-specific annotation.
+3. Pod-wide field.
+4. Pod-wide annotation.
```
copy/paste from seccomp? there is no pod-wide annotation

saschagrunert replied:
Yes, removed that one.

liggitt commented on keps/sig-node/24-apparmor-ga/README.md:
```diff
@@ -0,0 +1,682 @@
+
+The feature has been more or less unchanged ever since. Also note that the
+addition predates feature gates or our modern concept of feature lifecycle. So,
+even though the annotations include `beta` in the key, this is entirely useable
+on any production GA cluster.
```
this isn't accurate, the annotations are guarded by a feature gate and the feature can be disabled

saschagrunert replied:
Removed that section.

liggitt commented on keps/sig-node/24-apparmor-ga/README.md:
```diff
@@ -0,0 +1,682 @@
+
+- Declare AppArmor as GA
+- Fully document and formally spec the feature support
+- Add equivalent API fields to replace AppArmor annotations
```
This KEP goes beyond the annotation support in adding a pod-level profile field. Is that intentional?

saschagrunert replied:
Changed the line to include the pod wide field. Let's continue the discussion about the intention in https://github.com/kubernetes/enhancements/pull/1444#discussion_r503234404

liggitt commented on keps/sig-node/24-apparmor-ga/README.md:
```diff
@@ -0,0 +1,682 @@
+
+### Goals
+
+- Declare AppArmor as GA
```
Since it is currently beta and can be disabled via feature gate, a release in beta with the fields would make sense to me before promoting to GA.

saschagrunert replied:
Okay, what do you think about this proposal @tallclair @pjbgf ?

pjbgf replied:
I am just not sure whether is there much benefit on changing the API whilst keeping it in beta, considering this feature kept the same API "stable" in beta for a few years.
I feel that changing the API without calling it GA may send the message that we have doubts about this feature/API and more changes may come down the line.

Going down that route, would that start the count down for deprecating the previous syntax or would we support both syntaxes until the feature GAs?

liggitt commented on keps/sig-node/24-apparmor-ga/README.md:
```diff
@@ -0,0 +1,682 @@
+
+    // Load a profile defined on the node.
+    // The profile must be available on the node to work.
+    // The length of the profile is limited to 253 characters.
```
is this stricter than the current annotation validation (I think it is)? if so, that means currently working annotations could not be copied into the field

saschagrunert replied:
Is it about the profile length? That's the AppArmor limitation rather than any validation we do in Kubernetes. So I guess it's not stricter.

liggitt replied:
The question is what this API-level validation would do with existing deployment/replicasets/pods which have an annotation longer than 253 characters.

Current behavior:
- API operations are not blocked
- On nodes that don't support apparmor, those would run fine
- On nodes that do support apparmor, those would fail with a runtime error about the path being too long? (verify)

Proposed new behavior:
- translation between annotation and field in pod creation would fail on field validation?

saschagrunert replied:
If a profile name has more than 253 characters, then it would not possible to load it at all in memory. This means a runtime will fail to look it up. I think we should not enforce this by the API…should I remove the comment?

liggitt replied:
> This means a runtime will fail to look it up. I think we should not enforce this by the API…should I remove the comment?

It can be helpful to describe, but should not be confused with an API-enforced field limit

saschagrunert replied:
Alright, I removed it to avoid confusion. :+1:

liggitt commented on keps/sig-node/24-apparmor-ga/README.md:
```diff
@@ -0,0 +1,682 @@
+ensures that the fields are enforced even if the node version trails the API
+version (see [Version Skew Strategy](##version-skew-strategy))
+
+If _only_ AppArmor annotations are specified, copy the values into the
+corresponding fields. This ensures that existing applications continue to
```
an empty string annotation value or a `localhost/` annotation value is allowed (https://github.com/kubernetes/kubernetes/blob/deebe6619a1b4955b0cd109a8f05b1298a260310/pkg/security/apparmor/validate.go#L138-L142). what would happen with those?

saschagrunert replied:
CRI-O and containerd:
- apply the runtime default profile if the name is empty
- strip the `localhost/` prefix

liggitt replied:
add descriptions of those edge cases and what field values we will translate those to (and make sure the impl PR tests those)

saschagrunert replied:
Done, also added a note about the e2e tests.

saschagrunert commented on keps/sig-node/24-apparmor-ga/README.md:
```diff
@@ -0,0 +1,682 @@
+synchronization across nodes and 2) how difficult it can be to identify that
+profiles have been changed on disk/memory, after pods started using it. However,
+given its current "pseudo-GA" state, we don't want to change it with this KEP.
+Out of tree enhancements like the
+[seccomp-operator](https://github.com/kubernetes-sigs/seccomp-operator) can
```
Yes, we renamed the project to `security-profiles-operator`. I'll add a note that it might be used to sync profiles across nodes.

saschagrunert commented on keps/sig-node/24-apparmor-ga/README.md:
```diff
@@ -0,0 +1,682 @@
+The current support relies on profiles being loaded on all cluster nodes
+where the pods using them may be scheduled. It is also the cluster admin's
+responsibility to ensure the profiles are correctly saved and synchronised
+across the all nodes.
```
Yep, thanks! I add that.

saschagrunert commented on keps/sig-node/24-apparmor-ga/README.md:
```diff
@@ -0,0 +1,682 @@
+
+Prior to being marked GA, the feature tag will be removed from the AppArmor
+tests, and the tests will be migrated to the new fields API. Tests will be
+tagged as `[LinuxOnly]`.
```
I think we either could auto-skip those tests if a node does not support AppArmor or we leave them scoped to the feature. I think we should leave it as it is right now and scope them to `[Feature:AppArmor]`. Added a note about that.

saschagrunert commented on keps/sig-node/24-apparmor-ga/README.md:
```diff
@@ -0,0 +1,682 @@
+The AppArmor fields on a pod are immutable, which also applies to the
+[annotation](https://github.com/kubernetes/kubernetes/blob/b46612a74224b0871a97dae819f5fb3a1763d0b9/pkg/apis/core/validation/validation.go#L177-L182).
+
+Therefore, AppArmor annotation updates will be ignored. This maintains backwards
+API compatibility (no tightening validation), and makes a small stabilizing
+change to behavior (new Kubelets will ignore the update).
```
Yes, this section can be removed.

saschagrunert commented on keps/sig-node/24-apparmor-ga/README.md:
```diff
@@ -0,0 +1,682 @@
+The PodSecurityPolicy admission controller must continue to check the PSP object
+for annotations, as well as for fields.
+
+When setting default profiles, PSP only needs to set the field. The API
+machinery will handle setting the annotation as necessary.
```
Fixed as suggested :+1:

liggitt commented on keps/sig-node/24-apparmor-ga/README.md:
```diff
@@ -0,0 +1,679 @@
+but the tests are guarded by the `[Feature:AppArmor]` tag and not run in the
+standard test suites.
+
+Prior to being marked GA, the feature tag will be removed from the AppArmor
+tests, and the tests will be migrated to the new fields API. Tests will be
+tagged as `[Feature:AppArmor]` like it is implemented right now.
```
These two sentences seem to contradict each other

saschagrunert replied:
Oops, sorry. Fixed that one.

liggitt commented on keps/sig-node/24-apparmor-ga/README.md:
```diff
@@ -0,0 +1,679 @@
+
+- **How can this feature be enabled / disabled in a live cluster?**
+
+  The feature can be enabled/disabled by the `AppArmor` feature gate. This
+  behavior will not change when moving to GA.
```
```suggestion
  The feature can be enabled/disabled by the `AppArmor` feature gate. This
  feature gate can be used to disable the feature until it reaches GA.
```

liggitt commented on keps/sig-node/24-apparmor-ga/README.md:
```diff
@@ -0,0 +1,679 @@
+
+If no AppArmor annotations or fields are specified, no action is necessary.
+
+If the `AppArmor` feature is disabled per feature gate, then the annotations are
+cleared (current behavior) and the security context fields are being ignored.
```
```suggestion
If the `AppArmor` feature is disabled per feature gate, then the annotations and fields are cleared (current behavior).
```

saschagrunert commented:
@liggitt do you think we can move forward with this KEP?

liggitt commented on keps/sig-node/24-apparmor-ga/kep.yaml:
```diff
@@ -0,0 +1,46 @@
+participating-sigs:
+  - sig-api-machinery
+  - sig-auth
+status: implementable
```
shouldn't be "implementable" with "TBD" approvers and no PRR review :-)

saschagrunert replied:
Fixed :+1:

liggitt commented on keps/sig-node/24-apparmor-ga/kep.yaml:
```diff
@@ -0,0 +1,46 @@
+milestone:
+  alpha: "v1.4"
+  beta: "v1.4"
+  stable: "v1.20"
```
this missed 1.20 deadline, I'd expect 1.21

saschagrunert replied:
Yes, changed to 1.21

liggitt reviewed (commented):
happy to keep making progress on this. resolving the last few comments and switching stage to "provisional" would allow merging the bulk of this, then can chase a PRR review and approval

saschagrunert commented on keps/sig-node/24-apparmor-ga/README.md:
```diff
@@ -0,0 +1,681 @@
+If the `AppArmor` feature is disabled per feature gate, then the annotations and
+fields are cleared (current behavior).
+
+If _only_ AppArmor fields are specified, add the corresponding annotations. This
```
Alright, this should cover all use cases:
```
If a Pod with a container specifies an AppArmor profile by field/annotation,
then the container will only apply the Pod level field/annotation if no none
are set on the container level.
```

saschagrunert commented on keps/sig-node/24-apparmor-ga/README.md:
```diff
@@ -0,0 +1,681 @@
+
+### Version Skew Strategy
+
+Because the API is currently represented as (mutable) annotations, care must be
```
Ah, yeah I removed that paragraph to avoid confusion.

saschagrunert commented on keps/sig-node/24-apparmor-ga/README.md:
```diff
@@ -0,0 +1,681 @@
+
+### API
+
+The AppArmor API will be functionally equivalent to the current beta API. This
```
Yep, added a note about that.

saschagrunert commented:
> happy to keep making progress on this. resolving the last few comments and switching stage to "provisional" would allow merging the bulk of this, then can chase a PRR review and approval

Thank you for the detailed review! :pray: I think I addressed the review parts and will go for  a PRR review after this PR.

liggitt commented:
/lgtm
/approve

looks good for iteration, needs approvers and PRR review

saschagrunert commented:
/assign @dchen1107
Asking for aproval :)

saschagrunert commented:
@derekwaynecarr @dchen1107 do you think we can move forward with this KEP?

saschagrunert commented:
Friendly ping to @derekwaynecarr @dchen1107,  do you think we can move forward with this KEP? :innocent:

pjbgf commented:
Hey @derekwaynecarr @dchen1107 any chance you may have some time to take a look at this KEP so we can push it over the line?

ehashman commented:
/assign @derekwaynecarr

pacoxu commented:
/remove-lifecycle stale
Ping to @derekwaynecarr 😄

ehashman commented:
This currently isn't on the Node roadmap for 1.22: https://docs.google.com/document/d/1U10J0WwgWXkdYrqWGGvO8iH2HKeerQAlygnqgDgWv4E/edit#

saschagrunert commented:
/help 
I need someone who is willed to drive this to the end. My employer does not see this KEP as high priority any more and we should still get this done.

pjbgf commented:
@saschagrunert I agree we need to get this done, and I am happy to pick it up.

@ehashman is it OK to target this for `1.23`?

@derekwaynecarr do you mind taking a look at this PR please?

pjbgf commented:
/remove-help

saschagrunert commented:
> @saschagrunert I agree we need to get this done, and I am happy to pick it up.

Thank you! Updated the KEP metadata.

pjbgf commented:
/remove-lifecycle rotten
/help

Unfortunately I had no capacity in the last while to work on this. So it would be great if someone else could pick it up and move it forward.

MadhavJivrajani commented:
/remove-kind design
/kind feature
`kind/design` is migrated to `kind/feature`, see https://github.com/kubernetes/community/issues/6144 for more details

pjbgf commented:
/remove-lifecycle stale

PushkarJ commented:
I have added this as an agenda item for next week's SIG Security meeting to discuss and possibly find an owner to drive this. Will update the thread here after the discussion.

jan0ski commented:
> I have added this as an agenda item for next week's SIG Security meeting to discuss and possibly find an owner to drive this. Will update the thread here after the discussion.

Thanks @PushkarJ!  I'm happy to jump in and help drive this effort.

kikisdeliveryservice commented:
@jan0ski since https://github.com/kubernetes/enhancements/pull/3298 supersedes this PR could you close this one?

mccormickt commented:
/close
